torch/_subclasses/meta_utils.py

```diff
@@ -394,12 +398,48 @@ def is_c_of_r(complex_dtype, real_dtype):
 
                 else:
                     is_leaf = safe_is_leaf(t)
-                    sizes, strides, storage_offset = sym_sizes_strides_storage_offset(t)
-                    r = callback(
-                        lambda: torch.empty_strided(
-                            sizes, strides, dtype=t.dtype, device="meta"
-                        )
+                    sizes, strides, storage_offset = sym_sizes_strides_storage_offset(
+                        t, source
                     )
+
+                    # If we have a subclass that desugars into dense tensors,
+                    # perform our callback on each inner tensor.
+                    if is_traceable_wrapper_subclass(t):
+
+                        def empty_create(inner_t, inner_src):
+                            is_leaf = safe_is_leaf(inner_t)
+                            (
+                                inner_sizes,
+                                inner_strides,
+                                inner_storage_offset,
+                            ) = sym_sizes_strides_storage_offset(inner_t, inner_src)
+                            return torch.empty_strided(
+                                inner_sizes,
+                                inner_strides,
+                                dtype=inner_t.dtype,
+                                device="meta",
+                            )
```

Comment:
I tried this, and I realized that I can't due to the check [here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L423). I don't think I fully understand the purpose of this check (it's false if either we've seen the same storage before, or our strides/storage_offset are different). In the case of storage_offset, it looks like `r` is always created to have a storage offset of 0 ([code](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L400)), so we'll always go down this path if the original tensor has a non-zero storage offset. But I can't just update that check to be `r.stride() == t.stride() and r.storage_offset() == t.storage_offset()` - since this ends up adding unnecessary guards.